birdset/datamodule/embeddings_datamodule.py

```diff
@@ -0,0 +1,293 @@
+            test_data = Dataset.from_dict({key: [sample[key] for sample in test_data] for key in test_data[0]})
+
+            # Combine into a DatasetDict
+            dataset = DatasetDict({
+                'train': train_data,
+                'valid': val_data,
+                'test': test_data
+            })
+        else:
+            if self.val_batches == 0:
+                dataset['test'] = concatenate_datasets([dataset['valid'], dataset['test']])
+                dataset['valid'] = Dataset.from_dict({}) #! So that no split is created in the dataloader
+                
+            if self.low_train:
+                del dataset['train']
+                dataset['train'] = dataset['train_low']
+
+            del dataset['train_low']    
+                
+            
+            
+        return dataset
+
+    def _compute_embeddings(self, dataset):
+        """
+        Compute Embeddings for the entire dataset and store them in a new DatasetDict.
+        """
+        # Create a new DatasetDict to store the embeddings
+        embeddings_dataset = DatasetDict()
+        
+        #! For some reason the .map() from Datasets caused errors
+        # Iterate over each split in the dataset
+        for split in dataset.keys():
+            print(f">> Extracting Embeddings for {split} Split")
+            # Get the current split data
+            split_data = dataset[split]
+
+            # Create a list to store the embeddings for the current split
+            embeddings = []
+
+            # Iterate over each sample in the split
+            with torch.no_grad(): # No need to compute gradients
+                for sample in tqdm(split_data, total=len(split_data), desc="Extracting Embeddings"):
+                    # Get the embedding for the audio sample
+                    embedding = self._get_embedding(sample['audio'])
+                    
+                    # Add the embedding to the list
+                    sample['audio']['array'] = embedding.squeeze(0).cpu().numpy() # Move to CPU and convert to numpy
+                    embeddings.append(sample)
+
+                # Convert the list of embeddings to a tensor
+                #embeddings = torch.stack(embeddings)
+
+                # Create a new Dataset with the embeddings
+                if len(embeddings) > 0:
+                    embeddings_dataset[split] =  Dataset.from_dict({key: [sample[key] for sample in embeddings] for key in embeddings[0]})
+                else:
+                    embeddings_dataset[split] = Dataset.from_dict({'audio':[], 'labels':[]})    
+
+        return embeddings_dataset
+        
+
+    def _get_embedding(self, audio):
+        # Get waveform and sampling rate
+        waveform = torch.tensor(audio['array'], dtype=torch.float32).to(self.device) # Get waveform audio and move to GPU
+        dataset_sampling_rate = audio['sampling_rate']
+        # Resample audio
+        audio = self._resample_audio(waveform, dataset_sampling_rate)
```

Comment:
@XgamerTV Why resample to `dataset_sampling_rate` isn't it already in the sample rate of the dataset? Shouldn't we resample to the sample rate of the embedding model? Maybe we should add a function to the embedding model class `get_native_sample_rate`.

Reply:
Yes the function does that but is a bit misleading as the argument is the current sampling rate and the model's sampling rate is retrieved with self.sampling_rate